### DISCUSSION TITLE
Code precompiled with react compiler doesn't work in NextJS

### DISCUSSION THREAD

**Discussion by faradaytrs** [reactwg/react-compiler#46] [General Questions]
# Background

We have encountered a critical issue related to the Next.js React runtime compatibility after upgrading our monorepo to React 19. Despite following the officially recommended approach to precompile dependencies using React, the issue has left us unsure of the best way to handle it. https://github.com/vercel/next.js/discussions/73797

# Key Points:
1. We use a monorepo setup that includes:
* Several UI kit libraries precompiled with babel-plugin-react-compiler.
* A Next.js project that depends on these precompiled libraries.
2. React Compiler Usage:
* We previously relied on the react-compiler-runtime polyfill when using React 18, and everything worked fine.
* After upgrading to React 19 and removing the polyfill (since it is no longer supported), the Next.js project now fails to build due to runtime incompatibility.
3. Next.js Experimental Compiler:
* We intentionally avoided using the experimental React Compiler feature in Next.js to retain control over our precompiled dependencies.
	
# The Problem
1. Precompiling Dependencies as Recommended:
Following the official guidance, we precompiled our UI kit libraries using the React Compiler before building the Next.js project. However, after upgrading, build failures occur during prerendering.
2. Disabling prerendering results in runtime errors due to React runtime incompatibility.
Attempting the Next.js Experimental Compiler:
To investigate, we disabled the precompilation step for our dependencies and enabled the Next.js experimental React Compiler feature. Surprisingly, this approach worked, and the project built and ran without issues.
However, this is contrary to the recommended practices and does not align with how we intended to structure and optimize our setup.

# Conclusion

This situation has been a frustrating discovery, as it leaves us stuck between following official recommendations and a solution that technically works but contradicts best practices. We are eager to resolve this issue and would greatly appreciate clarity and guidance from the community or Next.js maintainers.

**Comment by poteto** [comment]
I'm not sure I'm following what the issue is exactly, could you provide a repro? If I'm understanding correctly, you still need to use the `react-compiler-runtime` dependency if you have precompiled those libraries, even if your app is on React 19. The polyfill will not be used (since React 19 natively has the compiler runtime APIs) but is still a necessary peer dependency from the pre-compiled libraries.

**Reply by faradaytrs** [reply]
I added it, but it's still the same problem, I will try my best to make a repro repo, but it will take time, is there anything i can provide to give you a quicker understanding? Maybe runtime error logs will be useful?

**Reply by poteto** [reply]
What is the problem exactly? What error?

**Reply by faradaytrs** [reply]
When i use precompiled library, I get errors like that:
<img width="660" alt="Screenshot 2024-12-16 at 11 07 41" src="https://github.com/user-attachments/assets/ea192897-666a-4101-aece-9918c8344058" />
And the next screenshot is a precompiled code file.
<img width="784" alt="Screenshot 2024-12-16 at 11 07 54" src="https://github.com/user-attachments/assets/1d986743-870b-4571-9f7e-a13714fd6b79" />
PeedDeps in the library:
<img width="488" alt="Screenshot 2024-12-16 at 11 09 54" src="https://github.com/user-attachments/assets/b04654be-39a2-4466-8236-2b359e5fe908" />
Please note that i'm using react-compiler-runtime peer.
<img width="571" alt="Screenshot 2024-12-16 at 11 10 55" src="https://github.com/user-attachments/assets/6aa2cdf8-e646-4e5e-a1c5-363234102ea5" />
Also, i have it in Next.js project.

**Reply by faradaytrs** [reply]
<img width="822" alt="Screenshot 2024-12-16 at 11 39 04" src="https://github.com/user-attachments/assets/d8649780-9f72-45ca-a203-774341ef06ab" />
<img width="724" alt="Screenshot 2024-12-16 at 11 39 17" src="https://github.com/user-attachments/assets/3a1f3292-cea6-4d23-8607-20d3d4edf6db" />

**Reply by faradaytrs** [reply]
I have no idea what is H and why it's undefined

**Reply by faradaytrs** [reply]
As i can see although i specify stable react version in dependencies, when i start nextjs it uses react-dom@19.0.0-rc-7283a213-20241206 renderer

**Reply by faradaytrs** [reply]
I found H here, not sure if it's the right "H"
<img width="975" alt="Screenshot 2024-12-16 at 12 40 22" src="https://github.com/user-attachments/assets/73e6d5cd-46c7-4023-898d-7fdefbb9b71a" />

**Reply by faradaytrs** [reply]
@poteto  here is the reproduction https://codesandbox.io/p/devbox/frosty-bird-xx8pz9
